android/src/main/java/org/whispersystems/signalservice/api/util/RealtimeSleepTimer.java

```diff
@@ -0,0 +1,82 @@
+package org.whispersystems.signalservice.api.util;
+
+import android.app.AlarmManager;
+import android.app.PendingIntent;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.os.Build;
+import android.os.SystemClock;
+import android.util.Log;
+
+import org.whispersystems.signalservice.api.util.SleepTimer;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A sleep timer that is based on elapsed realtime, so
+ * that it works properly, even in low-power sleep modes.
+ *
+ */
+public class RealtimeSleepTimer implements SleepTimer {
+  private static final String TAG = RealtimeSleepTimer.class.getSimpleName();
+
+  private final AlarmReceiver alarmReceiver;
+  private final Context context;
+
+  public RealtimeSleepTimer(Context context) {
+    this.context = context;
+    alarmReceiver = new RealtimeSleepTimer.AlarmReceiver();
+  }
+
+  @Override
+  public void sleep(long millis) throws InterruptedException {
+    context.registerReceiver(alarmReceiver,
+                             new IntentFilter(AlarmReceiver.WAKE_UP_THREAD_ACTION));
+
+    alarmReceiver.setAlarm(millis);
+
+    synchronized (this) {
+      wait(millis);
+    }
+
+    context.unregisterReceiver(alarmReceiver);
```

Comment:
If you throw InterruptedException from this method, then unregister needs to be in a finally block